### PR TITLE
[CALCITE-5881] Support to get foreign keys metadata in RelMetadataQuery

### PR DESCRIPTION
**Pull request by JingDas**
We can get constraints by RelOptTable#getReferentialConstraints method, but maybe can't get appropriate constraints at top relNode.
Foreign keys metadata is very useful in many optimize scenes. Such as it can be used in join eliminate when join type is inner join and some other star schema query optimize.
So support to get foreign keys metadata in RelMetadataQuery, it support to get composite foreign keys and  get the corresponding unique key on the arbitrary relNode if you want.
Tests are added to `RelMetadataTest`, the test code takes up a large part of the PR.

### CODE REVIEW

**Comment by LakeShen** [comment]
Hi @JingDas ,your pr's workflow has some abnormal information,please check it.
You could run this cmd in your local idea:
```sql
./gradlew check
```

File: core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+ * @see org.apache.calcite.plan.RelOptForeignKey
+ */
+public class RelOptForeignKey {
+
+  /** Foreign key and unique key relationships in bottom-up derivation. */
+  private final List<Pair<InferredConstraintKey, InferredConstraintKey>> constraints;
+  /** Position of the foreign key on the current {@link org.apache.calcite.rel.RelNode}. */
+  private final ImmutableBitSet foreignColumns;
+  /** Position of the unique key on the current {@link org.apache.calcite.rel.RelNode}. */
+  private final ImmutableBitSet uniqueColumns;
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:110]: Why is there information about a unique key in the RelOptForeignKey class?

File: core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java [hunk 2]
```diff
@@ -105,6 +106,43 @@ interface Handler extends MetadataHandler<UniqueKeys> {
...
+     * Foreign keys are represented as an {@link org.apache.calcite.util.ImmutableBitSet},
+     * where each bit position represents a 0-based output column ordinal.
+     *
+     * @param ignoreNulls if true, allow containing null values when determining
+     *                     whether the keys are foreign keys
+     *
+     * @return bit set of foreign keys(contains intermediate inferred foreign keys),
+     * or empty if not enough information is available to make that determination
+     */
+    Set<RelOptForeignKey> getForeignKeys(boolean ignoreNulls);
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java:128]: Is it @Nullable?

File: core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java [hunk 3]
```diff
@@ -105,6 +106,43 @@ interface Handler extends MetadataHandler<UniqueKeys> {
...
+
+    /**
+     * Handler API.
+     */
+    @FunctionalInterface
+    interface Handler extends MetadataHandler<ForeignKeys> {
+      Set<RelOptForeignKey> getForeignKeys(
+          RelNode rel,
+          RelMetadataQuery mq,
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java:137]: Is it @Nullable,like the method:
```sql
@Nullable Set<ImmutableBitSet> getUniqueKeys(RelNode r, RelMetadataQuery mq,
          boolean ignoreNulls);
```

File: core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java (hunk 1 above)
Comment (LakeShen) [reply core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:110]: I think a foreign key is a reference to the Primary key of another table,make uniqueColumns to referredPrimaryKeys.

Make the name reflect its true meaning

WDYT?

File: core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java (hunk 2 above)
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java:128]: Is it @Nullable ? Like this method:
```sql
@Nullable Set<ImmutableBitSet> getUniqueKeys(boolean ignoreNulls);

```

File: core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:40]: Good code comments are very important. Can you make your comments concise and clear so that others can understand them easily？

File: core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+   *   },
+   *   {
+   *     foreignColumns: {6, 7}
+   *     uniqueColumns: {8}
+   *   }
+   * ]
+   *
+   * @param foreignMapping foreignKey mapping relationship
+   * @param uniqueMapping uniqueKey mapping relationship
+   * @return mapped relOptForeignKey set
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:197]: What is it meaning?Give it a better description.

Comment (LakeShen) [reply core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:197]: Any other instructions about @return should have a good explanation

File: testkit/src/main/java/org/apache/calcite/test/RelMetadataFixture.java [hunk 6]
```diff
@@ -264,6 +265,18 @@ public RelMetadataFixture assertPercentageOriginalRows(Matcher<Double> matcher)
     return this;
   }
 
+  public RelMetadataFixture assertForeignKeys(Matcher<Set<RelOptForeignKey>> matcher,
+      Matcher<Set<RelOptForeignKey>> confirmedMatcher,
+      boolean ignoreNulls) {
+    RelNode rel = toRel();
+    RelMetadataQuery mq = rel.getCluster().getMetadataQuery();
+    Set<RelOptForeignKey> foreignKeys = mq.getForeignKeys(rel, ignoreNulls);
```
Comment (LakeShen) [inline testkit/src/main/java/org/apache/calcite/test/RelMetadataFixture.java:273]: You could write more test cases, just like you have getForeignKeys method in RelMdForeignKeys for different relNodes

File: core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+
+  public ImmutableBitSet getUniqueColumns() {
+    return uniqueColumns;
+  }
+
+  public List<Pair<InferredConstraintKey, InferredConstraintKey>> getConstraints() {
+    return constraints;
+  }
+
+  /** Returns the left side of a list of constraints. */
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:140]: What is meaning of left side,I think It should be described with clearer comments

File: core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+      List<Pair<InferredConstraintKey, InferredConstraintKey>> constraints) {
+    if (constraints.isEmpty()) {
+      return new ArrayList<>();
+    }
+    return constraints.stream()
+        .map(Pair::getKey)
+        .collect(Collectors.toList());
+  }
+
+  /** Returns the right side of a list of constraints. */
```
Comment (LakeShen) [inline core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:151]: same as above

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 9]
```diff
@@ -971,6 +989,1025 @@ final RelMetadataFixture sql(String sql) {
...
+                    InferredConstraintKey.of(DEPT_QNAME, 0, false))),
+            bitSetOf(1), ImmutableBitSet.of()));
+    Set<RelOptForeignKey> confirmedForeignKeys =
+        Sets.newHashSet(
+            RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(EMP_QNAME, 7, true),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, true))),
+            bitSetOf(1), bitSetOf(3)));
```
Comment (LakeShen) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:1364]: These two variables are used in almost every single test I see. They have too much duplicate code. You can define two methods that take different parameter types and then construct those two variables. This reduces code duplication.

**Comment by LakeShen** [comment]
Another point,Do you need add a class GeneratedMetadata_ForeignXXX like GeneratedMetadata_CollationHandler.java?
image

File: core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java (hunk 1 above)
Comment (JingDas) [reply core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java:110]: Add the referred prefix is more accessible, I will fix it.

File: core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java (hunk 3 above)
Comment (JingDas) [reply core/src/main/java/org/apache/calcite/rel/metadata/BuiltInMetadata.java:137]: It is reasonable to be consistent in behavior, get it.

**Comment by LakeShen** [comment]
Maybe we could close this PR,because we would add the functional dependency,the jira is https://issues.apache.org/jira/browse/CALCITE-5913

**Comment by LakeShen** [comment]
IMO,a foreign key is one of the table constraints, the current in Calcite, we could use Statistic# getReferentialConstraints method to get the table's foreign key.

So in an SQL plan(RelNode Tree),for example,if a RelNode is Project and its child is a TableScan,we want to use the foreign key information of TableScan. Instead of getting the foreign key information of Project, we should use functional dependencies.

In other words, only tables have foreign keys.The above is my understanding,If I understand this wrong,pls correct me,thanks.
